File: php/src/IpFilter.php
<?php

namespace FireWTWall;

/**
 * IP blacklist / whitelist filter with CIDR support (IPv4 and IPv6).
 */
class IpFilter
{
    private array $whitelist;
    private array $blacklist;

    public function __construct(array $whitelist = [], array $blacklist = [])
    {
        $this->whitelist = $whitelist;
        $this->blacklist = $blacklist;
    }

    /** Returns 'whitelist', 'blacklist', or null */
    public function check(string $ip): ?string
    {
        if (!empty($this->whitelist) && self::ipInList($ip, $this->whitelist)) {
            return 'whitelist';
        }
        if (!empty($this->blacklist) && self::ipInList($ip, $this->blacklist)) {
            return 'blacklist';
        }
        return null;
    }

    // ------------------------------------------------------------------ //
    // Static helpers (used by Request::resolveIp as well)
    // ------------------------------------------------------------------ //

    public static function ipInList(string $ip, array $list): bool
    {
        foreach ($list as $entry) {
            if (self::ipMatchesEntry($ip, $entry)) return true;
        }
        return false;
    }

    public static function ipMatchesEntry(string $ip, string $entry): bool
    {
        if (strpos($entry, '/') === false) {
            return $ip === $entry;
        }
        return self::ipInCidr($ip, $entry);
    }

    public static function ipInCidr(string $ip, string $cidr): bool
    {
        [$range, $prefix] = explode('/', $cidr, 2);
        $prefix = (int) $prefix;

        // IPv6
        if (filter_var($ip, FILTER_VALIDATE_IP, FILTER_FLAG_IPV6)) {
            return self::ipv6InCidr($ip, $range, $prefix);
        }
        // IPv4
        return self::ipv4InCidr($ip, $range, $prefix);
    }

    private static function ipv4InCidr(string $ip, string $range, int $prefix): bool
    {
        $ipLong    = ip2long($ip);
        $rangeLong = ip2long($range);
        if ($ipLong === false || $rangeLong === false) return false;
        $mask = $prefix === 0 ? 0 : (~0 << (32 - $prefix));
        return ($ipLong & $mask) === ($rangeLong & $mask);
    }

    private static function ipv6InCidr(string $ip, string $range, int $prefix): bool
    {
        $ipBin    = inet_pton($ip);
        $rangeBin = inet_pton($range);
        if ($ipBin === false || $rangeBin === false) return false;

        $byteCount  = (int) ceil($prefix / 8);
        $remaining  = $prefix % 8;

        for ($i = 0; $i < $byteCount - 1; $i++) {
            if ($ipBin[$i] !== $rangeBin[$i]) return false;
        }
        if ($byteCount > 0 && $remaining > 0) {
            $mask = 0xFF & (0xFF << (8 - $remaining));
            if ((ord($ipBin[$byteCount - 1]) & $mask) !== (ord($rangeBin[$byteCount - 1]) & $mask)) {
                return false;
            }
        }
        return true;
    }
}

File: aspnet/src/IpFilter.cs
using System;
using System.Net;
using System.Net.Sockets;

namespace FireWTWall
{
    /// <summary>
    /// IP blacklist / whitelist filter with CIDR support (IPv4 and IPv6).
    /// </summary>
    public sealed class IpFilter
    {
        private readonly string[] _whitelist;
        private readonly string[] _blacklist;

        public IpFilter(string[] whitelist, string[] blacklist)
        {
            _whitelist = whitelist ?? Array.Empty<string>();
            _blacklist = blacklist ?? Array.Empty<string>();
        }

        /// <summary>Returns "whitelist", "blacklist", or null.</summary>
        public string Check(string ip)
        {
            if (_whitelist.Length > 0 && IpInList(ip, _whitelist)) return "whitelist";
            if (_blacklist.Length > 0 && IpInList(ip, _blacklist)) return "blacklist";
            return null;
        }

        // ------------------------------------------------------------------ //
        // Static helpers (used by WafRequest as well)
        // ------------------------------------------------------------------ //

        public static bool IpInList(string ip, string[] list)
        {
            foreach (var entry in list)
            {
                if (IpMatchesEntry(ip, entry)) return true;
            }
            return false;
        }

        public static bool IpMatchesEntry(string ip, string entry)
        {
            if (!entry.Contains("/")) return ip == entry;
            return IpInCidr(ip, entry);
        }

        public static bool IpInCidr(string ip, string cidr)
        {
            var parts = cidr.Split('/');
            if (parts.Length != 2) return false;
            if (!int.TryParse(parts[1], out int prefix)) return false;

            string range = parts[0];

            IPAddress ipAddr, rangeAddr;
            if (!IPAddress.TryParse(ip, out ipAddr) || !IPAddress.TryParse(range, out rangeAddr))
                return false;

            if (ipAddr.AddressFamily != rangeAddr.AddressFamily)
                return false;

            if (ipAddr.AddressFamily == AddressFamily.InterNetwork)
                return Ipv4InCidr(ipAddr, rangeAddr, prefix);

            return Ipv6InCidr(ipAddr, rangeAddr, prefix);
        }

        private static bool Ipv4InCidr(IPAddress ip, IPAddress range, int prefix)
        {
            uint ipInt    = IpToUint(ip.GetAddressBytes());
            uint rangeInt = IpToUint(range.GetAddressBytes());
            uint mask     = prefix == 0 ? 0u : ~(0xFFFFFFFFu >> prefix);
            return (ipInt & mask) == (rangeInt & mask);
        }

        private static uint IpToUint(byte[] bytes)
        {
            return (uint)((bytes[0] << 24) | (bytes[1] << 16) | (bytes[2] << 8) | bytes[3]);
        }

        private static bool Ipv6InCidr(IPAddress ip, IPAddress range, int prefix)
        {
            byte[] ipBytes    = ip.GetAddressBytes();
            byte[] rangeBytes = range.GetAddressBytes();

            int fullBytes = prefix / 8;
            int remainder = prefix % 8;

            for (int i = 0; i < fullBytes; i++)
            {
                if (ipBytes[i] != rangeBytes[i]) return false;
            }

            if (remainder > 0 && fullBytes < ipBytes.Length)
            {
                byte mask = (byte)(0xFF << (8 - remainder));
                if ((ipBytes[fullBytes] & mask) != (rangeBytes[fullBytes] & mask))
                    return false;
            }

            return true;
        }
    }
}
